source/src/localisation/cs_TrackableParser.cpp

```diff
@@ -124,136 +223,371 @@ void TrackableParser::handleScannedDevice(scanned_device_t* device) {
+
+		// after deallocation, the filter is clearly gone.
+		parsingFilter = nullptr;
+	}
+
+	// check if the total size hasn't changed in the mean time
+	if (parsingFilter != nullptr && parsingFilter->runtimedata.totalSize != cmd_data->totalSize) {
+		// If total size to allocate changes something is really off on the host side
+		LOGe("Corrupt chunk of upload data received. Deallocating partially constructed filter.");
+		deallocateParsingFilter(parsingFilter->runtimedata.filterId);
+		return ERR_WRONG_STATE;
+	}
+
+	// check if we need to create a new filter
 	if (parsingFilter == nullptr) {
+		LOGTrackableParserDebug("parsing filter %d not found", cmd_data->filterId);
+		// command totalSize only includes the size of the cuckoo filter and its metadata, not the runtime data yet.
 		parsingFilter = allocateParsingFilter(cmd_data->filterId, cmd_data->totalSize);
 
 		if (parsingFilter == nullptr) {
 			// failed to handle command, no space.
+			LOGTrackableParserWarn("parsing filter allocation failed");
 			return ERR_NO_SPACE;
 		}
-
-		// initialize runtime data.
-		parsingFilter->runtimedata.filterId = cmd_data->filterId;
-		parsingFilter->runtimedata.crc      = 0;
-
-		// meta data and filter data will be memcpy'd from chunks,
-		// no need to copy those.
 	}
 
-	// WARNING(Arend): we're not doing corruption checks here yet.
-	// E.g.: when totalSize changes for a specific filterId before
-	//       a commit is reached, we have a parsingFilter allocated
-	//       but it's of the wrong size.
-	// Edit(Arend):
-	// 		As we keep a list of pointers _parsingFilters, as well
-	// 		as the end pointer to the byte array, this information
-	// 		is implicitly available even if multiple filters are
-	// 		uploaded in chunks concurrently. Just ptrdiff them.
+	// by now, the filter exists, is clean, and the incoming chunk is verified for totalSize consistency.
 
 	// chunk index starts counting from metadata onwards (ignoring runtimedata)
 	uint8_t* parsingFilterBase_ptr  = reinterpret_cast<uint8_t*>(&(parsingFilter->metadata));
 	uint8_t* parsingFilterChunk_ptr = parsingFilterBase_ptr + cmd_data->chunkStartIndex;
 
-	if (parsingFilterChunk_ptr + cmd_data->chunkSize > _filterBuffer + FILTER_BUFFER_SIZE) {
-		// chunk would overwrite end of filterbuffer.
-		// (Unlikely to happen, previous guards should have caught this.)
-		return ERR_NO_SPACE;
-	}
-
 	// apply filter chunk, counting chunk index from metadata onwards:
 	std::memcpy(parsingFilterChunk_ptr, cmd_data->chunk, cmd_data->chunkSize);
 
 	return ERR_SUCCESS;
 }
 
-cs_ret_code_t TrackableParser::handleRemoveFilterCommand(
-		trackable_parser_cmd_remove_filter_t* cmd_data) {
-	// TODO(Arend): implement later.
-	return ERR_NOT_IMPLEMENTED;
+cs_ret_code_t TrackableParser::handleRemoveFilterCommand(trackable_parser_cmd_remove_filter_t* cmd_data) {
+	startProgress();
+
+	return deallocateParsingFilter(cmd_data->filterId) ? ERR_SUCCESS : ERR_NOT_FOUND;
```

Comment:
When not found, isn't it a success as well?

Reply:
C++ segfaults when you delete a pointer twice. Seemed reasonable to inform the caller that they might be getting things wrong so that they can call getSummaries to check what's going on.

Reply:
change to ERR_SUCCESS_NO_CHANGE and document it.